README.md
# plugin.video.sickrage  
#### SickRage/SickBeard Interface Add-on for Kodi

View and control SickRage or SickBeard from within Kodi.

-----

## Kodi/XBMC Repo  
The repository for this add-on can be found here:  

https://github.com/Hiltronix/repo/

OR

https://hiltronix.com/xbmc/

-----

## API Parameters 

#### Add Show by TVDB ID:

    plugin://plugin.video.sickrage?action=addshow&tvdb_id=<tvdb_ID>

#### Add Show by Show Name at Keyboard Entry:

    - This is useful if you want to allow the user the option to edit the text first before searching.[/COLOR]
    plugin://plugin.video.sickrage?action=addshow&show_name=<url_encoded_text_for_the_show_name>

#### Add Show by Show Name and Skip Keyboard:

    plugin://plugin.video.sickrage?title=<url_encoded_text_for_the_exact_show_name_in_list>


-----

## Known Issues  

#### SickBeard Limitations & Issues:  
- Sickbeard does not (and devs won't add, last I checked) a function call to retrieve the API key, so hackish screen scraping is required to auto-retrieve the API Key.  
  This may break, so manual entry of API Key may be required.  
- The SickBeard API does not allow for retrieving the backlog list at this time.  

-----

## [Donations](https://hiltronix.com/donations/)  

If you would like to show your appreciation for this project, and/or help support it to keep it current, please consider making a donation.  
[Click Here](https://hiltronix.com/donations/).  
[![Support the project](https://github.com/Hiltronix/repo/blob/master/images/donate.png)](https://hiltronix.com/donations/)

-----

## Images  

[![Screen Show 1](https://github.com/Hiltronix/repo/blob/master/images/sickrage_addon_1sm.jpg)](https://github.com/Hiltronix/repo/blob/master/images/sickrage_addon_1.jpg)  
[![Screen Show 2](https://github.com/Hiltronix/repo/blob/master/images/sickrage_addon_2sm.jpg)](https://github.com/Hiltronix/repo/blob/master/images/sickrage_addon_2.jpg)  
[![Screen Show 3](https://github.com/Hiltronix/repo/blob/master/images/sickrage_addon_3sm.jpg)](https://github.com/Hiltronix/repo/blob/master/images/sickrage_addon_3.jpg)  
[![Screen Show 4](https://github.com/Hiltronix/repo/blob/master/images/sickrage_addon_4sm.jpg)](https://github.com/Hiltronix/repo/blob/master/images/sickrage_addon_4.jpg)  
[![Screen Show 5](https://github.com/Hiltronix/repo/blob/master/images/sickrage_addon_5sm.jpg)](https://github.com/Hiltronix/repo/blob/master/images/sickrage_addon_5.jpg)  
[![Screen Show 6](https://github.com/Hiltronix/repo/blob/master/images/sickrage_addon_6sm.jpg)](https://github.com/Hiltronix/repo/blob/master/images/sickrage_addon_6.jpg)  

-----

## Related Links

#### SickRage Website:  
https://www.sickrage.tv

#### SickRage Project on GitHub:  
https://github.com/SiCKRAGETV/SickRage

-----

## Note  

This add-on/plugin for Kodi/XBMC is an updated version of the Sickbeard plugin originally created by Zach Moore.  It expands on his original work starting at his version 1.0.10. before 2015.
http://forum.kodi.tv/showthread.php?tid=124808   

-----

## Disclaimer  

Kodi is a registered trademark of the XBMC Foundation. We are not connected to or in any other way affiliated with Kodi, Team Kodi, or the XBMC Foundation.  
Furthermore, any software, addons, or products offered by us will receive no support in official Kodi channels, including the Kodi forums and various social networks.  
We do not host any video content, nor do we condone piracy. Please use these add-ons at your own risk and abide by your country's copyright laws.


changelog.txt
[B]2017-07-18 v.3.0.3[/B]
[COLOR cyan]Fixed:[/COLOR] Bug fix in add show when unsuccessful.
[COLOR cyan]Fixed:[/COLOR] Add New Show duplicate show check was failing after last update.

[B]2017-07-06 v.3.0.2[/B]
[COLOR cyan]New:[/COLOR] You can now include "paused" shows in the Upcoming shows list.  Toggle switch is in Settings > Add-on Settings > Client Settings.

[B]2017-06-15 v.3.0.1[/B]
[COLOR cyan]Change:[/COLOR] Add New Show now queries theTVdb directly searching for possible show names.  Not using SickRage or SickBeard API.
[COLOR cyan]Fixed:[/COLOR] Error when adding a new show after show selection.

[B]2017-05-29 v.3.0.0[/B]
[COLOR cyan]New:[/COLOR] Clicking an episode in Episodes or History lists will start playing the show if it exists, otherwise it displays the video info dialog as before.
[COLOR cyan]New:[/COLOR] Added a service to auto update the TVdb meta data in the background while Kodi is running.  This allows for faster display of the show lists and show dialogs.
[COLOR cyan]New:[/COLOR] Background updating of local TVdb meta data cache for all "Upcoming" show episodes.
[COLOR cyan]New:[/COLOR] Background updating of local TVdb meta data cache for any new or missing episodes for all shows.
[COLOR cyan]New:[/COLOR] Background updating of local TVdb meta data cache for any local shows/episodes that have been changed on the TVdb service.
[COLOR cyan]New:[/COLOR] Background updating enabled by default.  Can be disabled, and change the timer interval, in Settings > Add-on Settings > Service.
[COLOR cyan]New:[/COLOR] Background updating default timer interval is 15 minutes, selectable from 5 to 1440 minutes (24 hours.)
[COLOR cyan]*** NOTE:[/COLOR] Internet data usage will be incurred on a regular basis to keep missing show data up to date while Kodi is running.
[COLOR cyan]New:[/COLOR] Added option to Settings menu to cache all missing episodes for all shows to the local TVdb meta data cache.  This is also done on every Kodi start.
[COLOR cyan]Change:[/COLOR] Re-arranged Settings menu list.
[COLOR cyan]Change:[/COLOR] Moved "Episodes List" to the top of the context menus.
[COLOR cyan]Change:[/COLOR] Small changes, dialog changes.
[COLOR cyan]Fixed:[/COLOR] Bug fixes.

[B]2017-04-30 v.2.1.5[/B]
- Fixed bug: "View Server Log File" was generating error message.

[B]2017-04-28 v.2.1.4[/B]
- Added anonymous version bump logging.
    - When version numbers change, the add-on name and version number are sent via a secure connection (SSL) to a private database, only accessible by the author, intended solely to determine install base.
    - No personal or identifiable info is transmitted, only version number and add-on name.
    - If server is unreachable, add-on will continue to function as normal.
    - Full disclosure, as the author is a privacy advocate.

[B]2017-04-26 v.2.1.3[/B]
- Updated About info.

[B]2017-04-24 v.2.1.2[/B]
- Updated About info.
- Added Settings option to add addon to favourites.

[B]2017-04-24 v.2.1.1[/B]
- Remove addon from the "Program add-ons" (executable) group.  Running the addon from this group stops video dialogs from opening, and meta data from being displayed in directory side bars.  This was done by a user Pull Request for convenience, it now just causes confusion on why things aren't working correctly.
- Added airs day of week, and airs time, to TV show video dialog status.
- Fixed some TV show meta data values.

[B]2017-04-15 v.2.1.0[/B]
- New context menu option: Search qBittorrent (if also installed.)
    - Sends the selected items show name, season and episode as a keyword search to the qBittorrent addon.
- Bug fixes.

[B]2017-04-15 v.2.0.9[/B]
- Changed some connection timeout values.
- Added busy indicator to Post Processing option.

[B]2017-04-14 v.2.0.8[/B]
- Added force manual post processing option to the Settings menu.

[B]2017-04-14 v.2.0.7[/B]
- Bug fixes and some code changes.

[B]2017-02-22 v.2.0.6[/B]
- Fix error where all lines in Show lists indicate "Paused" in error.

[B]2017-02-18 v.2.0.5[/B]
- The "Paused" show list no longer includes shows that have "Ended".
- You can open the TV show folder to play videos from the context menu.
    - This feature exits the Add-on, so you have to restart the addon to come back to where you were.
- Bug fixes.
- Cleaned up code.
- The link from "ExtendedInfo > Manage > Add to Sickbear" has been updated, so you may have to wait for an ExtendedInfo version bump before this works again.

[B]2017-02-16 v.2.0.4[/B]
- NOTE: Cached meta data may have to be re-downloaded and saved the first run, due to meta data version change requirements.
    - Re-caching meta data takes a few seconds per show.
- [B]Starting with this version, this app now only runs on Kodi 17 and higher[/B], due to using new API function calls.
- "Add New Show" feature now displays search results as video dialogs to preview before selecting, instead of having to guess from just a list of names.
- Added "Settings" menu item.
- Menu changes.
- Added images and artwork to seasons.
- Cleaned up code.
- Updated to "script.module.requests" version 2.12.4 to stay current.

[B]2017-02-13 v.2.0.3[/B]
- Minor changes.
- Improved error handling.

[B]2017-02-12 v.2.0.2[/B]
- Bug fixes.

[B]2017-02-11 v.2.0.1[/B]
- Improved list sorting.

[B]2017-02-08 v.2.0.0[/B]
- Switched to new TVdb scraper and meta data caching system.
- User experience greatly improved after TVdb data is cached locally.
- More meta data info is displayed.
- Manually update cached data directly from TVdb from context menu.
- Actors and their images are now displayed, when using Kodi 17.x or higher.
- Added banner image support.
- Total space used for cache now displayed in "Clear Cache".

[B]2017-02-05 v.1.4.3[/B]
- Fixed cancel dialog, return to main menu issue.

[B]2017-02-04 v.1.4.2[/B]
- Updated images.

[B]2017-02-04 v.1.4.1[/B]
- Updated images.

[B]2017-02-04 v.1.4.0[/B]
- Fix error retrieving API key using username and password.

default.py
import xbmc
import xbmcgui
import xbmcaddon
import xbmcplugin
import os
import sys
import urllib
import resources.lib.cache as cache
import resources.lib.common as common
import resources.lib.settings as settings
import resources.lib.upcoming as upcoming
import resources.lib.history as history
import resources.lib.seasons as seasons
import resources.lib.show_filter as show_filter
import resources.lib.shows as shows
import resources.lib.addTvdbShow as addshow
import resources.lib.episodes as episodes
import resources.lib.backlog as backlog
import resources.lib.log as log
import resources.lib.sickbeard as sickbeard
import resources.lib.TvdbApi as TvdbApi


# Initialize Sickbeard Class
Sickbeard = sickbeard.SB()


# Add the main directory folders.
def mainMenu():
    total_items = 6
    addDirectory('Upcoming - 1 Week', 2, True, settings.addon_path + '/upcoming.png', total_items)
    addDirectory('Upcoming - Extended', 22, True, settings.addon_path + '/upcomingplus.png', total_items)
    addDirectory('History', 3, True, settings.addon_path + '/history.png', total_items)
    if (settings.__servertype__ == "SickRage"):
        total_items += 1
        addDirectory('Backlog', 9, True, settings.addon_path + '/backlog.png', total_items)
    addDirectory('Show List', 1, True, settings.addon_path + '/manage.png', total_items)
    addDirectory('Add New Show', 7, False, settings.addon_path + '/add.png', total_items)
    addDirectory('Settings', 11, False, settings.addon_path + '/settings.png', total_items)
    xbmcplugin.endOfDirectory(int(sys.argv[1]))


# Add directory item.
def addDirectory(menu_item_name, menu_number, folder, icon, total_items):
    return_url = 'plugin://{}/?mode={}&name={}'.format(settings.pluginID, menu_number, urllib.quote_plus(menu_item_name))
    list_item = xbmcgui.ListItem(menu_item_name)
    list_item.setArt({'icon': icon, 'thumb': icon, 'poster': icon, 'fanart': '', 'banner': '', 'clearart': '', 'clearlogo': '', 'landscape': ''})
    xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=return_url, listitem=list_item, isFolder=folder, totalItems=total_items)


# Get the parameters from the URL supplied as an arg to this script.
def getParameters():
    param=[]
    try:
      paramstring=sys.argv[2]
      if len(paramstring)>=2:
              params=sys.argv[2]
              cleanedparams=params.replace('?','')
              if (params[len(params)-1]=='/'):
                      params=params[0:len(params)-2]
              pairsofparams=cleanedparams.split('&')
              param={}
              for i in range(len(pairsofparams)):
                      splitparams={}
                      splitparams=pairsofparams[i].split('=')
                      if (len(splitparams))==2:
                              param[splitparams[0]]=splitparams[1]
      return param
    except:
      return param
          
          
# Initialize URL parameters.
url = None
name = None
number = None
menu_number = None
action = None
show_name = None
show_id = None
tvdb_id = None
season = 0
episode = 0
params = getParameters()

# Parse internal URL.
try:
    url = urllib.unquote_plus(params["url"])
    print url
except:
    pass

try:
    name = urllib.unquote_plus(params["name"])
    print name
except:
    pass

try:
    number = urllib.unquote_plus(params["number"])
    print number
except:
    pass

try:
    menu_number = int(params["mode"])
    print menu_number
except:
    pass
    
try:
    show_name = urllib.unquote_plus(params["title"])
    menu_number = 7
    print show_name
except:
    pass

try:
    action = urllib.unquote_plus(params["action"])
    menu_number = 8
    print action
except:
    pass

try:
    show_name = urllib.unquote_plus(params["show_name"]).decode('utf-8')
    print show_name
except:
    pass

try:
    tvdb_id = urllib.unquote_plus(params["tvdb_id"])
    print tvdb_id
except:
    pass

try:
    tvdb_id = urllib.unquote_plus(params["show_id"])    # Used by ExtendedInfo.
    print tvdb_id
except:
    pass

try:
    season = urllib.unquote_plus(params["season"])
    print season
except:
    pass

try:
    episode = urllib.unquote_plus(params["episode"])
    print episode
except:
    pass

# Open directories based on selection.
if menu_number == None:
    mainMenu()
       
elif menu_number == 1:  # Show list type [All, Continuing, Ended, Paused].
    show_filter.menu(sys.argv[1])
        
elif menu_number == 14: # Shows list.
    shows.menu(handle=sys.argv[1], filter=name)
        
elif menu_number == 2:  # Upcoming 1 Week.
    upcoming.menu(sys.argv[1], False)
        
elif menu_number == 22:   # Upcoming Extended.
    upcoming.menu(sys.argv[1], True)
        
elif menu_number == 3:  # History list.
    history.menu(sys.argv[1])

elif menu_number == 4:  # Seasons.
    seasons.menu(sys.argv[1], tvdb_id, show_name)

elif menu_number == 5:  # Episodes.
    episodes.menu(sys.argv[1], tvdb_id, show_name, number)

elif menu_number == 6:  # Open the video dialog window.
    xbmc.executebuiltin('XBMC.Action(Info)')
    
elif menu_number == 7:  # Add a new show.
    addshow.AddShow(show_name)

elif menu_number == 8:  # Action values.
    if action == 'addshow':  # Add a new show based on show name.  Used by the Manage section of ExtendedInfo.
        if tvdb_id:
            # Use tvdb_id to get show_name to pass as argument for user confirmation.
            try:
                xbmc.executebuiltin("ActivateWindow(busydialog)")
                show_name = TvdbApi.GetNameFromTvdb(tvdb_id)
            finally:
                xbmc.executebuiltin("Dialog.Close(busydialog)")
            addshow.AddShowDetails(tvdb_id, show_name)
        else:
            common.messageWindow('Missing Parameter', 'Add new show request received,[CR]but the TVdb ID was not included.')

elif menu_number == 9:  # Backlog list.
    backlog.menu(sys.argv[1])

# Settings menu.
elif menu_number == 11:
    ret = 0
    while ret >= 0:
        dialog = xbmcgui.Dialog()
        ret = dialog.select("Settings", ["Add-on Settings", "Run Post Processing", "Add Add-on to Favouties", "Check if All Episodes are Cached", "Clear Cache", "Show Server Version", "View Server Log File", "Change Log", "About"])
    
        if ret == 0:    # App Settings (Kodi settings.)
            xbmc.executebuiltin('XBMC.Addon.OpenSettings({0})'.format(settings.pluginID))
            ret = -1    # Exit Settings menu so Kodi App Settings can be on top.
    
        if ret == 1:    # Run Post Processing.
            try:
                xbmc.executebuiltin("ActivateWindow(busydialog)")
                msg, res = Sickbeard.PostProcessing()
            finally:
                xbmc.executebuiltin("Dialog.Close(busydialog)")
            if res:
                common.messageWindow('Post Processing', 'Msg: {}[CR]Result: {}'.format(msg, res))
    
        if ret == 2:    # Add App to Favourites.
            try:
                xml_open = '<favourites>\n'
                xml_fav = '    <favourite name="{0}" thumb="special://home/addons/{1}/icon.png">ActivateWindow(10025,&quot;plugin://{1}/&quot;,return)</favourite>\n'.format(settings.pluginName, settings.pluginID)
                xml_close = '</favourites>\n'
                xbmc.executebuiltin("ActivateWindow(busydialog)")
                filename = xbmc.translatePath('special://home/userdata/favourites.xml')
                if os.path.isfile(filename):
                    with open(filename, 'r') as f:
                        data = f.read()
                    # Check is fav already exists.  If not, add it to end of file.
                    find_str = 'name="{}"'.format(settings.pluginName)
                    if find_str not in data:
                        data = data.replace(xml_close, xml_fav + xml_close)
                else:
                    data = xml_open + xml_fav + xml_close
                # Save favourites file.
                f = open(filename, 'w')
                f.write(data)
                f.close()
                common.messageWindow('Add to Favourites', 'Done.')
            finally:
                xbmc.executebuiltin("Dialog.Close(busydialog)")
    
        if ret == 3:    # Check Cache for All Episodes.
            try:
                if settings.my_addon.getSetting('TVdbUpdateAllEpisodesEnabled') == 'true':
                    if settings.my_addon.getSetting('TVdbUpdateAllEpisodesRunning') == 'false':
                        state = '[COLOR cyan]Scheduled[/COLOR]'
                    else:
                        state = '[COLOR gold]Active[/COLOR]'
                    if common.DialogYesNo('Check Cache For All Episodes', text='Operation: {}[CR]Progress: [COLOR gold]{}%[/COLOR][CR][CR][COLOR grey](Re-open this dialog to see updated progress.)[/COLOR]'.format(state, settings.my_addon.getSetting('TVdbUpdateAllEpisodesProgress')), noLabel='Close', yesLabel='Cancel Operation'):
                        # Set flags to turn service operation off.
                        settings.my_addon.setSetting('TVdbUpdateAllEpisodesEnabled', 'false')
                        settings.my_addon.setSetting('TVdbUpdateAllEpisodesRunning', 'false')
                        settings.my_addon.setSetting('TVdbUpdateAllEpisodesProgress', '0')
                else:
                    if common.DialogYesNo('Check Cache For All Episodes  [Not Running]', u'{0} This background operation may take some time to complete.[CR]{0} If Kodi is shutdown before completion, you will have to manually select this setting again.[CR]{0} You can check progress by revisiting this selection.'.format(u'[COLOR cyan]\u2022[/COLOR]'), noLabel='Close', yesLabel='Start'):
                        # Set flags to turn service operation on at next interval.
                        settings.my_addon.setSetting('TVdbUpdateAllEpisodesEnabled', 'true')
                        settings.my_addon.setSetting('TVdbUpdateAllEpisodesRunning', 'false')
                        settings.my_addon.setSetting('TVdbUpdateAllEpisodesProgress', '0')
            finally:
                pass
                
    
        if ret == 4:    # Clear Cache.
            try:
                xbmc.executebuiltin("ActivateWindow(busydialog)")
                size = common.GetDirSizeFormatted(settings.cache_dir)
            finally:
                xbmc.executebuiltin("Dialog.Close(busydialog)")
            if common.DialogYesNo('Clear Cached Images and Meta Data?', u'Cache storage is using: [COLOR gold]{0}[/COLOR]'.format(size), noLabel='Close', yesLabel='Clear Cache'):
            #if common.selectNoYes('Clear cached images and meta data?  [{0}]'.format(size), 'No', 'Yes') == 1:
                try:
                    xbmc.executebuiltin("ActivateWindow(busydialog)")
                    cache.ClearImages()
                    cache.ClearMetaData()
                finally:
                    xbmc.executebuiltin("Dialog.Close(busydialog)")
                common.CreateNotification(header='Image Cache', message='Cleared', icon=xbmcgui.NOTIFICATION_INFO, time=5000, sound=False)
    
        if ret == 5:    # SickRage/SickBeard Version.
            api, version = Sickbeard.GetVersion()
            common.messageWindow('Server Version', 'API Version: {0}[CR]Version: {1}'.format(api, version))
    
        if ret == 6:    # View log files.
            log.main()
    
        if ret == 7:    # Change log.
            try:
                xbmc.executebuiltin("ActivateWindow(busydialog)")
                filename = os.path.join(settings.addon_path, 'changelog.txt')
                if os.path.isfile(filename):
                    with open(filename, 'r') as f:
                        data = f.read()
                else:
                    data = 'Change log not available.'
            finally:
                xbmc.executebuiltin("Dialog.Close(busydialog)")
            w = common.TextViewer_Dialog('DialogTextViewer.xml', settings.addon_path, header='Change Log', text=data)
            w.doModal()
    
        if ret == 8:    # About.
            try:
                xbmc.executebuiltin("ActivateWindow(busydialog)")
                filename = os.path.join(settings.addon_path, 'about.txt')
                if os.path.isfile(filename):
                    with open(filename, 'r') as f:
                        data = f.read()
                else:
                    data = 'About file not available.'
                data = '[COLOR gold]{} v.{}[CR]by {}[/COLOR][CR][COLOR cyan]{}[CR]{}[/COLOR][CR][CR]'.format(settings.pluginName, settings.pluginVersion, settings.pluginAuthor, settings.pluginSummary, settings.pluginDesc) + data
            finally:
                xbmc.executebuiltin("Dialog.Close(busydialog)")
            w = common.TextViewer_Dialog('DialogTextViewer.xml', settings.addon_path, header='About', text=data)
            w.doModal()
    

# Update a show's images.
elif menu_number == 13:
    try:
        xbmc.executebuiltin("ActivateWindow(busydialog)")
        # Update Poster image.
        Sickbeard.GetShowPoster(tvdb_id, update=True)
        # Update Fanart image.
        Sickbeard.GetShowFanArt(tvdb_id, update=True)
    finally:
        xbmc.executebuiltin("Dialog.Close(busydialog)")
    common.CreateNotification(header='Updated Cached Images for', message=show_name, icon=xbmcgui.NOTIFICATION_INFO, time=5000, sound=False)
    xbmc.executebuiltin('Container.Refresh')

# Open the TV show folder (to select a video to watch.)
elif menu_number == 15:
    try:
        data = common.JsonRpc('libTvShows', 'VideoLibrary.GetTVShows', {"properties": ["imdbnumber", "title", "year"], "sort": {"order": "ascending", "method": "title"}})
        shows = common.getFromDict(data,['result', 'tvshows'], [])
        tvshowid = None
        for show in shows:
            if show.get('imdbnumber') == str(tvdb_id):
                tvshowid = show.get('tvshowid')
                break
        if tvshowid:
            xbmc.executebuiltin('ActivateWindow(videos,videodb://tvshows/titles/{0}/, True)'.format(tvshowid))
    except Exception, e:
        common.messageWindow('Open Show Folder Error', str(e))
        print e
    
